### ISSUE TITLE
support entity kind filter / allow list at entity providers ingesting Location entities

### ISSUE THREAD

**Issue by pjungermann** [backstage/backstage#12880]
## Feature Suggestion

We replaced many processor with entity providers which ingest Location entities for discovered catalog files.

At processors, you would configure a static location within the app config which then gets picked up by it.
Additionally, you could make use of the feature to add the "allow" rule with a set of allowed entity kinds to be ingested.

This is not possible anymore with the new entity providers.

## Context

- Allow to provide a similar config at all these providers.
- Pass the rules config to the locations created.
- Maybe an abstract class, etc. which those providers could extend instead of writing it again for each of them.

**Comment by namco1992** [comment]
Hi @pjungermann,

We recently encountered the same issue, do you have a plan in mind about how to implement this?

Another thing that bothers me is that the entities from providers are not restrained by the `catalog.rules` neither. the rules only apply to `referredEntities` emitted by processors, and the entities ingested by providers don't fall in that category. 

Maybe one can argue that those rules only apply to things that are included in the locations mechanism, but IMO since the name of the config is `catalog.rules`, I was expecting it's a set of rules that guard the entrance to catalog, regardless of where it's come from. WDYT @pjungermann?

**Comment by Rugvip** [comment]
Yep this is a bit of a gap. I think perhaps the most straightforward short-term fix for this would be to make it possible to include locations in `catalog.rules`? But then again, wouldn't you just add that location to `catalog.locations` at that point instead? I would actually be curious to hear what use-cases you have in mind for more granular rules of entities discovered by providers.

The fact that rules aren't applied to entity providers is by design, as entity providers are generally trusted. This does mean that you can write an insecure provider though, and I think there is room for adding more ways to help guard against that.

**Comment by pjungermann** [comment]
@namco for me this is only about providers which ingest Location entities.

For other kind of providers, they should provide you with filter
capabilities if this makes sense. E.g., at the msgraph provider, you can
filter which Users, Groups, etc. will be added. The kinds are fixed though.

With Locations, this is different. You could include any kind of entity
there.

Usually, you have a dedicated provider for org data, ingestion of
User/Group/... entities.

So.. what is missing here? What is different to static location entities
defined at the configuration file?

Generally, catalog.rules[].allow is applied globally to control which kinds
are allowed.

At static location entities, you can override the allowed kinds per
location.
Hence, you can exclude Group and User globally but allow it for a specific
Location entity.

This mechanism is not available for non-static Location entities ingested
by a provider. Here, only the global rule will apply.

Let's assume, you have a provider which ingests Location entities which
include User and Group records. However, you only want to allow it for
Locations entities by this provider.
This is not possible and you will need to allow them globally. Therefore,
anyone could add User and Group in, let's say, any discovered repository.

An idea I have so far is to add this option to the provider config and add
the overridden rules to the ingested Location entities at the provider.
Maybe this can be abstracted into a shared/base implementation (many
providers are quite aligned at the moment).

Then you could define your override at the provider instance level.

Patrik Oldsberg ***@***.***> schrieb am Do., 1. Sept. 2022,
12:59:

**Comment by Rugvip** [comment]
Tbh it feels a little bit risky to combine discovery providers and rule overrides that apply to all locations it emits.

Wanna come back to whether this is a problem you've faced in practice or not? I'm feeling that static `catalog.locations` with rules should be filling this gap in practice, and that any solution based on entity providers is likely to have some security pitfalls.

**Comment by pjungermann** [comment]
@Rugvip we have such a use case with Groups provided via the S3 provider, but general repos via the VCS-related provider shouldn't be allowed to add Groups. And at the moment, we have such global rules.

And statically defining all the files instead of using the provider is not an option.

I don't see much difference to catalog.locations having their own overrides, too.

**Comment by jhaals** [comment]
@pjungermann do you have suggestions for how we could configure rules per entity provider?

**Comment by pjungermann** [comment]
At the static locations, you have something like the following at the moment:

```yaml
catalog:
  rules:
    - allow:
        - Component
        - API
        - Resource
        - Template
        - System
        - Domain
        - Location
  locations:
    - type: file
      target: ../catalog-model/examples/acme-corp.yaml
      rules:
        - allow: [User, Group]
```

and now we have

```yaml
catalog:
  rules:
    - allow:
        - Component
        - API
        - Resource
        - Template
        - System
        - Domain
        - Location
  providers:
    awsS3:
      org-data:
        bucketName: acme-org
    # ...other providers
```

which we could extend to:

```yaml
catalog:
  # ...
  providers:
    awsS3:
      org-data:
        bucketName: acme-org
        rules:
          - allow: [User, Group]
    # ...other providers
```

But I didn't look much into the current implementation and how to make use of it there.

**Comment by pjungermann** [comment]
So, the `DefaultCatalogRulesEnforcer` loads both options `catalog.rules` and `catalog.locations.[].rules` directly from the config. There is no association with the Location entity created from the config.

The check `isAllowed(entity: Entity, location: LocationSpec): boolean;` only has access to the entity to be checked (e.g., Group, User) and its data incl. annotations like `backstage.io/managed-by-location` or `backstage.io/managed-by-origin-location` as well as the LocationSpec from the Location which was responsible for attempting to add it.

Rules are either global or scoped to 1..n LocationSpecs (target + type).
For the non-global ones, rules will only be checked if one of the locations (specs) matches the entity's LocationSpec.

Fetching the config from the providers here would not help as we have no LocationSpecs we could generate out of it - esp. without internal knowledge about the Location.

If we had access to the locationKey we might be able to solve this though.
```typescript
locationKey: this.getProviderName(),
```

```typescript
  getProviderName(): string {
    return `awsS3-provider:${this.config.id}`;
  }
```
(where config.id is `org-data` in the example -- same logic for all providers and could be standardized/enforced more)

This information is actually available at the DefaultCatalogProcessingOrchestrator:
```typescript
      for (const deferredEntity of collectorResults.deferredEntities) {
        if (
          !this.options.rulesEnforcer.isAllowed(
            deferredEntity.entity,
            context.originLocation,
          )
        ) {
```

as part of the `DeferredEntity`
```typescript
export type DeferredEntity = {
  entity: Entity;
  locationKey?: string;
};
```

So, if we change the snippet above to
```typescript
      for (const deferredEntity of collectorResults.deferredEntities) {
        if (
          !this.options.rulesEnforcer.isAllowed(
            deferredEntity,
            context.originLocation,
          )
        ) {
```
and the interface to 
```typescript
export type CatalogRulesEnforcer = {
  isAllowed(entity: DeferredEntity | Entity, location: LocationSpec): boolean;
};
```
(actually, we would rather support both for backwards compatibility reasons and handle it as before if only Entity was passed)

Of course, we would also need to know what the used location key (or provider name which is used for it always so far) actually is. `getProviderName()` is part of the low-level interface `EntityProvider`.

Originally, I thought about adding a shared base class for all Location-adding providers like the AwsS3 and all SCM-related providers. Probably, we could use this base class to read the config and provide it in a way that it provides rules + locationKey to the rule enforcer implementation.


Those are just some ad-hoc thoughts. 🙂

**Comment by pjungermann** [comment]
PS: I would not support this for other types of providers like msgraph, etc. which directly ingest the target kinds. These should provide enough (filter) capabilities to exclude what you don't want to have.

**Comment by namco1992** [comment]
Hi @pjungermann,

Thanks for the clarification! However, I still have some concerns here.

Firstly, I understood you and @Rugvip's position that the `catalog.rules` shouldn't be applied to providers that directly ingest the target kinds. But I think we owe some explanations to the users. One can [dig into the code base and find out that it's only for `deferredEntities`](https://github.com/backstage/backstage/blob/6de25bee18bd3e60b2000f3b364a1a48fbd15c30/plugins/catalog-backend/src/processing/DefaultCatalogProcessingOrchestrator.ts#L138-L141), and then trace back to providers and processors to understand the difference between directly ingested entities and `deferredEntities` that are typically derived from a location entity. 

At least to me it's not very intuitive that something called `catalog.rules` are not applied to all entities in catalog, but just "some" entities ingested in a specific way. So my question is: 
- If it's just for location entities, then doesn't `catalog.location.rules` make more sense?
- If it's by design, then we should state it clearly in the documentation.

Secondly, I don't see why having a global catalog rule for all entities (including the directly injected ones) is a bad thing. We have three different models now: 
- The discovery processors, emit entities that are filtered by location rules
- The entity providers, emit entities that are not filtered by location rules currently **but maybe later**
- Other providers, directly ingest the targeted kinds and are not filtered by location rules

IMO it's a significant cognitive load. I can hardly imagine explaining this to other engineers without like 5 diagrams and an hour-long session, and half of them might leave the room with even more questions 😂 . I really appreciate the effort to fill the gap between the processors and providers, but I also believe that the entity filter doesn't have to be this complicated.

**Comment by pjungermann** [comment]
> and the interface to
> 
> ```ts
> export type CatalogRulesEnforcer = {
>   isAllowed(entity: DeferredEntity | Entity, location: LocationSpec): boolean;
> };
> ```
> 
> (actually, we would rather support both for backwards compatibility reasons and handle it as before if only Entity was passed)

Instead of adding/allowing `DeferredEntity` we could add the `locationKey` as optional argument instead so potential users of this interface (e.g., for customizations) wouldn't need to bother with the difference between Entity and DeferredEntity.

```ts
export type CatalogRulesEnforcer = {
  isAllowed(entity:Entity, location: LocationSpec, locationKey?: string): boolean;
};
```

**Comment by pjungermann** [comment]
@namco1992 

> At least to me it's not very intuitive that something called `catalog.rules` are not applied to all entities in catalog, but just "some" entities ingested in a specific way. So my question is:
> 
> * If it's just for location entities, then doesn't `catalog.location.rules` make more sense?
> * If it's by design, then we should state it clearly in the documentation.

I get where you are coming from.

I'm not one of the maintainers, so cannot really reply. I assume that catalog.rules was supposed to be extensible with different kind of rules. It receives an array of rule configs and only supports the rule "allow" which acts as an allow set for entity kinds.

Also, there is already catalog.locations which accepts an array of location configs which themselves can have their own rules config (override).

catalog.location.rules and catalog.locations might be a bit confusing.

Re docs: https://backstage.io/docs/features/software-catalog/configuration#catalog-rules
> By default the catalog will only allow ingestion of entities with the kind Component, API and Location. In order to allow entities of other kinds to be added, you need to add rules to the catalog. Rules are added either in a separate catalog.rules key, or added to statically configured locations. [...]

I guess you are right that the current docs don't really mention that this only applies to entities ingested via Locations.

Maybe you want to provide a PR to improve this?

**Comment by pjungermann** [comment]
> Secondly, I don't see why having a global catalog rule for all entities (including the directly injected ones) is a bad thing. We have three different models now:
> 
> * The discovery processors, emit entities that are filtered by location rules
> * The entity providers, emit entities that are not filtered by location rules currently **but maybe later**
> * Other providers, directly ingest the targeted kinds and are not filtered by location rules

I don't think it makes sense to apply the global kind allow list to entities ingested by providers.

The goal of the allow list is to have control over entities ingested where others can define any kind of entity. This is the case at Locations / catalog-info.yaml files. If I add i.e. the msgraph provider it is clear that it will ingest User and Group kind entities from the start. The allow list would not help in any way to improve control here. Rather the provider itself needs to provide more fine-grained control to be able to manage which User and/or Group entities will be added from the total set.

Hence, for me it only about entities which originate from a Location entity. And for these, the global rules already apply. But as I wrote here before, you have no ability to allow a different set within the scope of a provider as you are able within the scope of a Location added to the config.

And there is no difference between processors and providers (re you list above). Both have the same "weakness" of *only* adhering to the global rule without any option to alter it within their scope.

Current state: 

- global rules: apply to all entities originating from a Location entity
- Location-specific rule (override): only for Locations defined in the config (static Locations)

**Comment by freben** [comment]
Hm, this is all getting a bit tricky. Let's see if I understood things correctly.

Would it be a (basic, but perhaps functional) solution if we had something like this that matched on the originating URL annotation?

```yaml
catalog:
  rules:
    - allow: [Component, API]
    - allowOriginUrlPatterns:
        'gs://org-exports-.*': [Group, User]
```

This won't guard the immediate emission surface from the providers themselves (they can emit anything into the refresh table), but it would be a simple way of using an updated version of the existing rules enforcer mechanism to perhaps be able to at least avoid adding sensitive kinds to the "global" rules.

Let me know if I misunderstood somewhere along the way.

**Comment by pjungermann** [comment]
@freben I guess this could cover a big chunk of use cases incl. ours.

As long as you can represent all Location targets / URLs with 1+ patterns without overlap, it will be good enough. But that is likely quite advanced.

And yes, for me it is only about Location emitting providers and what these might contain. (e.g., awsS3, bitbucketCloud, github, etc.)

**Comment by namco1992** [comment]
> I guess you are right that the current docs don't really mention that this only applies to entities ingested via Locations.
Maybe you want to provide a PR to improve this?

Yea I would love to if the maintainers agree too.

And what @freben proposed makes sense to me too, altho I think I will still reserve my opinion on the entity filter 😳 , I agree that it would be good if we can fix it for location emitting providers for now.

**Comment by freben** [comment]
Yeah the thing with entity providers is, they are often (but not always!) made to emit exactly a specific thing. Like the org ingestion ones, they are utterly _unable_ to emit anything else than `Group` and `User` entities, by the way their code is written, so it'd be unnecessary extra work to have to also allowlist those kinds for that provider in your config. Same with most discovery-like things that emit `Location` entities, it's all they do. So a kind based rule makes little sense there too, for the "initial layer".

**Comment by pjungermann** [comment]
I assume that this issue was resolved with the merged PR #14584 (part of the upcoming release 1.10.0)

It should enable orgs to set up location allow rules for URL patterns. It keeps it entirely decoupled from the provider/creator of the location entity (== independent of entity provider implementations).

Feel free to comment here if you see something missing.
